man/systemd.resource-control.xml

```diff
@@ -264,8 +264,14 @@
         <term><varname>MemoryMin=<replaceable>bytes</replaceable></varname></term>
 
         <listitem>
-          <para>Specify the memory usage protection of the executed processes in this unit. If the memory usages of
-          this unit and all its ancestors are below their minimum boundaries, this unit's memory won't be reclaimed.</para>
+          <para>Specify the memory usage protection of the executed processes in this unit.
+          When reclaiming memory, the unit is treated as if it was using less memory resulting in memory
+          to be preferentially reclaimed from unprotected units.
+          This is a strong protection, the OOM killer will be invoked if there is no other reclaimable memory.
+          A <varname>MemoryLow=</varname> allocation also needs to be assigned to ancestor slice units
```

Comment:
I don't grok this sentence.

Reply:
I wanted point out that you must pass a `MemoryLow=` allocation to all parent slice units for the protection to work. i.e. basically answer the question "If is set 'MemoryLow=' on a unit, will this just work or may I need to do more than that? (answer: yes, be careful/aware of allocations on all parents)".

Reply:
IIUC the documentation correctly, any ancestor cgroups and their descendants without memory.low set will not be protected. But that doesn't prevent the cgroup with protections set from being protected. If we required MemoryLow= to be set all the way up that'd be super annoying to use. In addition, memory.low cannot be set on the root cgroup, so it's not possible to do this on all ancestors anyway.

Reply:
Well, the root cgroup is special, you don't need to set anything there.

But for all the other ones, yeah, you need to set it in-between. For example, the `MemoryLow=` allocation of `system.slice` will be distributed across all children. If it is `50M` and two children each set `50M` for themselves, then each child will only receive `25M`.

Unless of course, systemd would instead automatically set `100M` as the `memory.low` allocation of `system.slice`.

Reply:
I [tried to explain](https://lore.kernel.org/lkml/20200729140537.13345-1-mkoutny@suse.com/T/#m31840a33676e07ddfc4a033d2d461d49e96e8d6f) how this behaves in the kernel docs. The protection needn't be set on all ancestors generally, but on all ancestors up to the reclaim target (typically it's same as root but it's not so if per-cgroup limits are used). That being said, I'm not sure systemd docs should go into such a detail, perhaps just point out the top-down "allocation" necessity + reference to the kernel docs.

BTW this paragraph should have 's/MemoryLow/MemoryMin/'. For simplification, I'd merge the two paragraphs under one `MemoryMin=, MemoryLow=` header and explained in the end that `MemoryMin=` protection may trigger OOM.

Reply:
I like the idea of merging the two sections, so went ahead to do that.

I changed the sentence in question again. However, I am starting to think we need something smarter on the systemd side, in particular if one considers corner cases like implicit slice units for template services.

Maybe a solution would be a new feature along the lines of `MaxMemoryLow=` effectively setting `MemoryLow=min(own.MaxMemoryLow, sum(child.MemoryLow) + own.MemoryLow)`. But even that has somewhat odd semantics, as it could make sense to default to `max` for all slices except immediate children of `-.slice`.